coderd/database/modelmethods.go

```diff
@@ -235,6 +241,199 @@ func (s APIKeyScopes) Name() rbac.RoleIdentifier {
 	return rbac.RoleIdentifier{Name: "scopes[" + strings.Join(names, "+") + "]"}
 }
 
+// APIKeyScopeSet merges expanded scopes with the API key's DB allow_list. If
+// the DB allow_list is a wildcard or empty, the merged scope's allow list is
+// unchanged. Otherwise, the DB allow_list overrides the merged AllowIDList to
+// enforce the token's resource scoping consistently across all permissions.
+type APIKeyScopeSet struct {
+	Scopes    APIKeyScopes
+	AllowList AllowList
+}
+
+var _ rbac.ExpandableScope = APIKeyScopeSet{}
+
+func (s APIKeyScopeSet) Name() rbac.RoleIdentifier { return s.Scopes.Name() }
+
+func (s APIKeyScopeSet) Expand() (rbac.Scope, error) {
+	merged, err := s.Scopes.expandRBACScope()
+	if err != nil {
+		return rbac.Scope{}, err
+	}
+	merged.AllowIDList = mergeAllowLists(merged.AllowIDList, s.AllowList)
+	return merged, nil
+}
+
+// ScopeSet returns the scopes combined with the database allow list. It is the
+// canonical way to expose an API key's effective scope for authorization.
+func (k APIKey) ScopeSet() APIKeyScopeSet {
+	return APIKeyScopeSet{
+		Scopes:    k.Scopes,
+		AllowList: k.AllowList,
+	}
+}
+
+// mergeAllowLists combines the allow list produced by RBAC expansion with the
+// API key's stored allow list. The result enforces both constraints: any
+// resource must be allowed by the scope *and* the database filter. Wildcards in
+// either list are respected and short-circuit appropriately.
+//
+// Intuition: scope definitions provide the *ceiling* of what a key could touch,
+// while the DB allow list can only narrow that set. A few illustrative cases:
+//
+//	Scope AllowList			DB AllowList								-> Result
+//	[*:*]								[workspace:A]								-> [workspace:A]
+//	[workspace:*]				[workspace:A, workspace:B]	-> [workspace:A, workspace:B]
+//	[workspace:A]				[workspace:A, workspace:B]	-> [workspace:A]
+//	[] (legacy scope)		[workspace:A]								-> [workspace:A]
+//
+// Today most API key scopes expand with an empty allow list (meaning "no
+// scope-level restriction"), so the merge simply mirrors what the database
+// stored. Only scopes that intentionally embed resource filters would trim the
+// DB entries.
+func mergeAllowLists(scopeList []rbac.AllowListElement, dbList AllowList) []rbac.AllowListElement {
+	// Empty DB list means no additional restriction.
+	if len(dbList) == 0 {
+		// Defensive: API keys should always persist a wildcard allow list, but
+		// treating an empty slice as "no additional filter" preserves legacy data.
+		return scopeList
+	}
```

Comment:
Doesn't our migration ensure this is never the case? The db always has a wildcard entry